clang/lib/Sema/SemaTemplateInstantiateDecl.cpp

```diff
@@ -574,28 +574,37 @@ static void instantiateIntelSYCTripleLFunctionAttr(
   S.addIntelTripleArgAttr<AttrName>(New, *Attr, XDimExpr, YDimExpr, ZDimExpr);
 }
 
-template <typename AttrName>
-static void instantiateIntelFPGAMemoryAttr(
+static void instantiateIntelFPGAForcePow2DepthAttr(
     Sema &S, const MultiLevelTemplateArgumentList &TemplateArgs,
-    const AttrName *Attr, Decl *New) {
+    const IntelFPGAForcePow2DepthAttr *Attr, Decl *New) {
   EnterExpressionEvaluationContext Unevaluated(
       S, Sema::ExpressionEvaluationContext::ConstantEvaluated);
   ExprResult Result = S.SubstExpr(Attr->getValue(), TemplateArgs);
-  if (!Result.isInvalid()) {
-    if (std::is_same<AttrName, IntelFPGABankWidthAttr>::value ||
-        std::is_same<AttrName, IntelFPGANumBanksAttr>::value)
-      return S.AddOneConstantPowerTwoValueAttr<AttrName>(New, *Attr,
-                                                         Result.getAs<Expr>());
+  if (!Result.isInvalid())
+    return S.AddOneConstantValueAttr<IntelFPGAForcePow2DepthAttr>(
+        New, *Attr, Result.getAs<Expr>());
+}
 
-    if (std::is_same<AttrName, IntelFPGAPrivateCopiesAttr>::value ||
-        std::is_same<AttrName, IntelFPGAMaxReplicatesAttr>::value)
-      return S.addIntelSingleArgAttr<AttrName>(New, *Attr,
-                                               Result.getAs<Expr>());
+static void instantiateIntelFPGABankWidthAttr(
+    Sema &S, const MultiLevelTemplateArgumentList &TemplateArgs,
+    const IntelFPGABankWidthAttr *Attr, Decl *New) {
+  EnterExpressionEvaluationContext Unevaluated(
+      S, Sema::ExpressionEvaluationContext::ConstantEvaluated);
+  ExprResult Result = S.SubstExpr(Attr->getValue(), TemplateArgs);
+  if (!Result.isInvalid())
+    S.AddOneConstantPowerTwoValueAttr<IntelFPGABankWidthAttr>(
+        New, *Attr, Result.getAs<Expr>());
+}
 
-    if (std::is_same<AttrName, IntelFPGAForcePow2DepthAttr>::value)
-      return S.AddOneConstantValueAttr<AttrName>(New, *Attr,
-                                                 Result.getAs<Expr>());
-  }
+static void instantiateIntelFPGANumBanksAttr(
+    Sema &S, const MultiLevelTemplateArgumentList &TemplateArgs,
+    const IntelFPGANumBanksAttr *Attr, Decl *New) {
+  EnterExpressionEvaluationContext Unevaluated(
+      S, Sema::ExpressionEvaluationContext::ConstantEvaluated);
+  ExprResult Result = S.SubstExpr(Attr->getValue(), TemplateArgs);
+  if (!Result.isInvalid())
+    S.AddOneConstantPowerTwoValueAttr<IntelFPGANumBanksAttr>(
+        New, *Attr, Result.getAs<Expr>());
```

Comment:
both are doing same. You can combine them to one. Is there a reason to have a separate instantiation function.

Reply:
I noticed open-source clang usually kept it separate and so I did as well.

Reply:
Please keep them separate for now as I'm expecting that to make future work a bit easier.

Reply:
> I noticed open-source clang usually kept it separate and so I did as well.

Thanks for the info and clarification.